File: backend/app/schemas/task.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, field_validator

from app.models.task import TaskStatus
from app.schemas.topic import TopicResponse


class TaskCreateRequest(BaseModel):
    title: str
    description: str | None = None
    due_date: datetime | None = None
    topic_ids: list[uuid.UUID] = []

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Title must not be empty")
        if len(v) > 255:
            raise ValueError("Title must not exceed 255 characters")
        return v


class TaskUpdateRequest(BaseModel):
    title: str | None = None
    description: str | None = None
    due_date: datetime | None = None
    topic_ids: list[uuid.UUID] | None = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            if not v:
                raise ValueError("Title must not be empty")
            if len(v) > 255:
                raise ValueError("Title must not exceed 255 characters")
        return v


class TaskStatusUpdateRequest(BaseModel):
    status: TaskStatus
    result_note: str | None = None


class TaskOrderUpdateRequest(BaseModel):
    manual_order: int


class TaskResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    description: str | None
    due_date: datetime | None
    status: TaskStatus
    result_note: str | None
    archived: bool
    done_at: datetime | None
    archived_at: datetime | None
    manual_order: int | None
    topics: list[TopicResponse]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TaskFilterParams(BaseModel):
    window: Literal["today", "3days", "week", "all"] | None = None
    topic_id: uuid.UUID | None = None
    q: str | None = None
    page: int = 1
    limit: int = 20

    @field_validator("limit")
    @classmethod
    def limit_max(cls, v: int) -> int:
        if v > 30:
            return 30
        if v < 1:
            return 1
        return v

    @field_validator("page")
    @classmethod
    def page_min(cls, v: int) -> int:
        if v < 1:
            return 1
        return v

File: frontend/src/types/task.ts
import type { Topic } from './topic'

export type TaskStatus = 'todo' | 'in_progress' | 'done'

export interface Subtask {
  id: string
  task_id: string
  title: string
  status: TaskStatus
  sort_order: number
  created_at: string
  updated_at: string
}

export interface SubtaskCreatePayload {
  title: string
}

export interface SubtaskUpdatePayload {
  title?: string
  status?: TaskStatus
  sort_order?: number
}

export interface Task {
  id: string
  title: string
  description: string | null
  due_date: string | null
  status: TaskStatus
  result_note: string | null
  archived: boolean
  done_at: string | null
  archived_at: string | null
  manual_order: number | null
  created_at: string
  updated_at: string
  topics: Topic[]
  recurring_template_id: string | null
  subtask_count: number
  subtask_done_count: number
  subtasks: Subtask[]
}

export interface TaskCreatePayload {
  title: string
  description?: string | null
  due_date?: string | null
  status?: TaskStatus
  topic_ids?: string[]
}

export interface TaskUpdatePayload {
  title?: string
  description?: string | null
  due_date?: string | null
  result_note?: string | null
  topic_ids?: string[]
}

export interface TaskStatusUpdatePayload {
  status: TaskStatus
  result_note?: string | null
}

export interface TaskOrderUpdatePayload {
  manual_order: number
}

export type TaskFilterWindow = 'today' | '3days' | 'week' | 'all'

export interface TaskFilterParams {
  window?: TaskFilterWindow
  topic_id?: string
  q?: string
  page?: number
  limit?: number
}
